core/include/sched.h

```diff
@@ -278,6 +278,63 @@ typedef void (*sched_callback_t)(kernel_pid_t active, kernel_pid_t next);
 void sched_register_cb(sched_callback_t callback);
 #endif /* MODULE_SCHED_CB */
 
+#if (IS_USED(MODULE_SCHED_RUNQ_CALLBACK)) || defined(DOXYGEN)
+/**
+ * @brief   Scheduler runqueue callback
+ *          function has to be provided by the user of this API
+ *          it will be called: when the scheduler is run,
+ *                             when a thread enters the active queue or
+ *                             when the last thread leaves a queue
+ *
+ * @param   prio      the priority of the runqueue that changed
+ *
+ * @warning This API is not intended for out of tree users.
+ *          Breaking API changes will be done without notice and
+ *          without deprecation. Consider yourself warned!
+ */
+extern void sched_runq_callback(uint8_t prio);
+#endif
+
+/**
+ * @brief   round robin a runqueue
+ *
+ * @param   prio      the priority of the runqueue to round robin
+ *
+ * @warning This API is not intended for out of tree users.
+ *          Breaking API changes will be done without notice and
+ *          without deprecation. Consider yourself warned!
+ */
+static inline void sched_round_robin_runqueue(uint8_t prio)
+{
+    clist_lpoprpush(&sched_runqueues[prio]);
+}
+
+/**
+ * @brief   count the number of threads in a runqueue up to 2
+ *
+ * @param[in]   prio      the priority of the runqueue to get the length from
+ * @returns     number of threads in a runqueue up to 2
+ * @retval 0    The runqueue is empty
+ * @retval 1    There is exactly one thread in the runqueue
+ * @retval 2    There are at least two threads in the runqueue
+ *
+ * @warning This API is not intended for out of tree users.
+ *          Breaking API changes will be done without notice and
+ *          without deprecation. Consider yourself warned!
+ */
+static inline uint_fast8_t sched_runqueue_len2(uint8_t prio)
```

Comment:
this thing is threre to avoid the O(n) complexity by limiting n thus making its formal complexity O(1)

i will check if "count upto n"  is optimized the same way.

Reply:
there's no "count upto n" yet

Reply:
a first test showed that a simple capped variant of `clist_count_upto(` generates more code with -Og (native).
since i also still would need some thing like:
```
static inline uint_fast8_t sched_runqueue_len2_bob(uint8_t prio)
{
    return clist_count_upto(&sched_runqueues[prio],2);
}
```
i stop my test right now 
i did test with an L073 -both variants generated the same code size
so if clist is in need for this feature i will port it move it and would use the sched_rq_len2 you see above this

Reply:
Note that what you are actually testing for in your code `clist_is_empty()`, `clist_contains_exactly_one()` and `clist_contains_more_than_one()`.

I personally would just add those three as `static inline` functions. All of them should be smaller than the overhead of a function call in terms of both ROM size in CPU cycles for most platforms, so there is no to ROM win in having one function handling all cases. And I personally consider `clist_contains_exactly_one(list)` easier to read than `clist_count_upto(list, 2) == 1`.